mitodl/mitxonline
# Remove program field from course model

# What are the relevant tickets?
https://github.com/mitodl/mitxonline/issues/1683
https://github.com/mitodl/hq/issues/1663
https://github.com/mitodl/mitxonline/issues/1464
https://github.com/mitodl/mitxonline/pull/1480
https://github.com/mitodl/mitxonline/issues/1220
https://github.com/mitodl/mitxonline/issues/1681

# Description
Removes the program FK field from the Course model.  The Program/Course relationship is derived from the requirements and electives defined on the Program instance.

There is a new program relationship table created which is a link table for associating 2 programs together.  By associating 2 programs together in this link table, financial aid approvals in one course will be applied to the linked program's courses.

# How can this be tested?
Non-Related Program

 Can create program, tiers, courseware page, and financial assistance request form
 Can create a course as a requirement in the program, with products and an about page
 Can create a course as an elective in the program, with products and an about page
 "Financial assistance" CtAs appear in the appropriate spots - in the program dashboard cards, in the upsell cards, etc.
 Can submit and approve a financial assistance request for the program
 Personalized pricing displays appropriately on the course about pages
 Can enroll in the course programs
 After enrolling, the program appears successfully in the dashboard and displays all associated courses within the program
 Upgrading course enrollment results in the proper financial assistance tier being applied as discount
Related Programs, No Overlapping Courses

 Can create program, tiers, courseware page, and financial assistance request form
 Can create a course as a requirement in the program, with products and an about page
 Can create a course as an elective in the program, with products and an about page
 "Financial assistance" CtAs appear in the appropriate spots - in the program dashboard cards, in the upsell cards, etc.
 Can submit and approve a financial assistance request for the program
 Personalized pricing displays appropriately on the course about pages
 Can enroll in the course programs
 After enrolling, the program appears successfully in the dashboard and displays all associated courses within the program
 Upgrading course enrollment results in the proper financial assistance tier being applied as discount
 Can add a second program to the system, with a relation to the first, and complete the first three steps for it
 Personalized pricing appears in the courses assigned to the second program
 Can create a secondary financial assistance request page for the second program
 Attempting to apply for financial assistance should be disallowed
This delivered me to the approved page for the proper program (test program 2 rather than 1)
 The dashboard should automatically see both related programs
 Enrolling in a course in the second program should result in a program enrollment for the second program; one should not exist prior to this
Notes: I only enrolled in the course in the second program for https://github.com/mitodl/mitxonline/issues/16 - no upgrade. I didn't see the upsell card on the course details page for the other program I enrolled but did not upgrade in. I do see the upsell card on the dashboard page.

Related Programs, Overlapping Courses

 Can create program, tiers, courseware page, and financial assistance request form
 Can create a course as a requirement in the program, with products and an about page
 Can create a course as an elective in the program, with products and an about page
 "Financial assistance" CtAs appear in the appropriate spots - in the program dashboard cards, in the upsell cards, etc.
 Can add a second program to the system, with a relation to the first
 Can create a course as a requirement in the second program, with products and an about page
 Can add the new course to the first program as either a requirement or an elective
 Enrolling in the overlapping course should result in a program enrollment for both programs
 After clearing program enrollments, enrolling in the non-overlapping courses should also result in a program enrollment in both
 After applying for financial assistance in the first program, personalized pricing should be shown in both program's courses
Tiers Check

 After applying for financial assistance for a program with related programs, adjust one program so that the applied tier is better in one than the other. The tier with the better value should apply.
Should be noted that the request for program 1 approved me with the tiers for program 1 too, so even when checking out with a course only in program 2, I got the tiers for the first program.
